itest/lnd_estimate_route_fee_test.go

```diff
@@ -0,0 +1,226 @@
+			FeeBaseMsat:               1_000,
+			FeeProportionalMillionths: 1,
+			ChanId:                    421337,
+		}
+
+		bobExpensiveBaseFeeMsat = int64(2_000)
+		bobExpensiveFeeRatePpm  = int64(2_000)
+		bobExpensiveHopHint     = &lnrpc.HopHint{
+			NodeId:                    mts.bob.PubKeyStr,
+			FeeBaseMsat:               uint32(bobExpensiveBaseFeeMsat),
+			FeeProportionalMillionths: uint32(bobExpensiveFeeRatePpm),
+			ChanId:                    815,
+		}
+
+		singleRouteHint = []*lnrpc.RouteHint{
+			{
+				HopHints: []*lnrpc.HopHint{
+					bobHopHint,
+				},
+			},
+		}
+		singleLastHopFeeMsat      = probeAmountMsat*1/1_000_000 + 1_000
+		singleFeeToBobMsat        = (singleLastHopFeeMsat+probeAmountMsat)/1_000_000 + 1_000
+		expectedSingleHintFeeMsat = singleFeeToBobMsat + singleLastHopFeeMsat
+
+		lspDifferentFeesHints = []*lnrpc.RouteHint{
+			{
+				HopHints: []*lnrpc.HopHint{
+					bobHopHint,
+				},
+			},
+			{
+				HopHints: []*lnrpc.HopHint{
+					bobExpensiveHopHint,
+				},
+			},
+		}
+		lspDiffLastHopFeeMsat           = probeAmountMsat*bobExpensiveFeeRatePpm/1_000_000 + bobExpensiveBaseFeeMsat
+		lspDiffFeeToBobMsat             = (lspDiffLastHopFeeMsat+probeAmountMsat)/1_000_000 + 1000
+		lspDiffExpectedMultiHintFeeMsat = lspDiffFeeToBobMsat + lspDiffLastHopFeeMsat
+	)
+
+	//nolint:lll
+	var testCases = []*estimateRouteFeeTestCase{
+		// the cheapest route charges...
+		// 1000 msat base fee + 100 msat(1ppm*100_000sat)
+		{
+			name:                    "graph based estimate",
+			probeInitiator:          mts.alice,
+			destination:             mts.bob,
+			amount:                  probeAmount,
+			expectedRoutingFeesMsat: int64(1100),
+			expectedFailureReasons:  lnrpc.PaymentFailureReason_FAILURE_REASON_NONE,
+		},
+		// single hop payment is free.
+		{
+			name:                    "graph based estimate, 0 fee",
+			probeInitiator:          mts.alice,
+			destination:             mts.dave,
+			amount:                  probeAmount,
+			expectedRoutingFeesMsat: int64(0),
+			expectedFailureReasons:  lnrpc.PaymentFailureReason_FAILURE_REASON_NONE,
+		},
+		{
+			name: "probe based estimate, single" +
+				" route hint",
+			probeInitiator:          mts.alice,
+			destination:             mts.paula,
+			amount:                  probeAmount,
+			routeHints:              singleRouteHint,
+			timeoutSeconds:          10,
+			expectedRoutingFeesMsat: expectedSingleHintFeeMsat,
+			expectedFailureReasons:  lnrpc.PaymentFailureReason_FAILURE_REASON_NONE,
+		},
+		{
+			name: "probe based estimate, " +
+				"multiple route hints, diff lsp fees",
+			probeInitiator:          mts.alice,
+			destination:             mts.paula,
+			amount:                  probeAmount,
+			routeHints:              lspDifferentFeesHints,
+			timeoutSeconds:          10,
+			expectedRoutingFeesMsat: lspDiffExpectedMultiHintFeeMsat,
+			expectedFailureReasons:  lnrpc.PaymentFailureReason_FAILURE_REASON_NONE,
+		},
+	}
```

Comment:
given that this work enables probing to LSPs, a test case where the private channel doesn't actually exist feels necessary. That would guarantee that it actually probes up to the LSP.

Reply:
@champo 
To clarify, the assumption is that LSPs block payment probes. So if our heuristic `isLsp` turns true we can't really tell if a private channel that is indicated in the hop hints actually exists or not. We will probe to the LSP regardless and manually determine the last hop's fee.
The test cases above already ensure that we are probing up to the Lsp(Bob) because the routes over Dave and Eve are high fee, so we expect the payment probe to go through Carol to Bob, which actually is the case since it matches our expected fees.

Let me know if you have further doubts or test suggestions here. Thanks for your review !

Reply:
I'm getting a bit confused from your answer, so I'll recap what I understand from the harness and these test cases so it's easier to find out where the confusion is coming from:
1. There's an open private channel between Bob and Paula
2. Bob is the LSP
3. The tests expect the probe to be done up to Bob

Given all these if the code had a bug and decide to probe up to Paula, wouldn't this test case succeed too? Paula would be the one to return the incorrect payment details error which would be accepted as a sucessful probe.

Reply:
That makes sense, I got you. I added a test case with a node without any edges and use it as a destination of the probe, with a route hint through the previously used LSP. This should return the same results as the case where the private hop exists.